File: app/src/main/java/com/hamohdy/whatsappdirect/CountrySelectorActivity.java
package com.hamohdy.whatsappdirect;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.WindowInsetsControllerCompat;

import com.hamohdy.whatsappdirect.CountryListAdapter.CountrySelectedListener;
import com.hamohdy.whatsappdirect.databinding.ActivityCountrySelectBinding;

/**This activity displays a list of countries to pick a country code form, and also shows the device default country at the top.*/
public class CountrySelectorActivity extends AppCompatActivity implements CountrySelectedListener {

    private final CountryListAdapter adapter = new CountryListAdapter(this);
    private ActivityCountrySelectBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.WhatsAppDirect);
        super.onCreate(savedInstanceState);

        binding = ActivityCountrySelectBinding.inflate(getLayoutInflater());
        View root = binding.getRoot();
        setContentView(root);

        //We need dark status bar text during the day
        if (!getResources().getBoolean(R.bool.nightMode)) {
            WindowInsetsControllerCompat controller = new WindowInsetsControllerCompat(getWindow(), root);
            controller.setAppearanceLightStatusBars(true);
        } else {
            root.setBackgroundColor(Color.BLACK);
            getWindow().setStatusBarColor(Color.BLACK);
        }

        binding.counter.setText(Constants.size);
        binding.recycler.post(() -> {
            binding.recycler.setAdapter(adapter);
            binding.scroller.attachRecyclerView(binding.recycler);
            binding.scroller.setSectionIndexer(adapter);
        });
        getDeviceCountry();
    }

    @Override
    public void countrySelected(int position) {
        Intent intent = new Intent();
        if (position != -1) intent.putExtra("country", Constants.countries.get(position));
        setResult(RESULT_OK, intent);
        finish();
    }


    /**Since the main activity will always be launched first, we are relying on the launch intent to get the device default country. If it is not passed, just remove visibility to prevent bad experience.*/
    private void getDeviceCountry() {

        Country deviceCountry = getIntent().getParcelableExtra("default");
        if (deviceCountry == null) {
            binding.deviceDefault.getRoot().setVisibility(View.GONE);
            binding.defaultHeader.setVisibility(View.GONE);
            return;
        }

        String name = deviceCountry.name + ", " + deviceCountry.isoCode;
        String dialingCode = deviceCountry.isdCode;
        int referenceId = deviceCountry.flagCode;

        binding.deviceDefault.countryName.setText(name);
        binding.deviceDefault.countryCode.setText(dialingCode);
        binding.deviceDefault.flag.setImageResource(referenceId);

        binding.deviceDefault.getRoot().setOnClickListener(v -> countrySelected(getIntent().getIntExtra("position", -1)));
    }

}

File: app/src/main/java/com/hamohdy/whatsappdirect/CountrySelectorActivity.kt
package com.hamohdy.whatsappdirect

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.hamohdy.whatsappdirect.databinding.ActivityCountrySelectBinding

/**This activity displays a list of countries to pick a country code form, and also shows the device default country at the top. */
class CountrySelectorActivity : AppCompatActivity() {

    //view binding
    private lateinit var binding: ActivityCountrySelectBinding

    //countries adapter. Passing a function to the parameter for handling item click.
    private val adapter = CountriesAdapter{ countrySelected(it) }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.WhatsAppDirect)
        super.onCreate(savedInstanceState)

        binding = ActivityCountrySelectBinding.inflate(layoutInflater)
        val root: View = binding.root
        setContentView(root)

        //We need dark status bar text during the day
        if (resources.getBoolean(R.bool.nightMode)) {
            root.setBackgroundColor(Color.BLACK)
            window.statusBarColor = Color.BLACK
        }

        //setting up layout display
        binding.counter.text = countries.size.toString()
        binding.recycler.adapter = adapter
        binding.scroller.attachRecyclerView(binding.recycler)
        binding.scroller.setSectionIndexer(adapter)

        //setting up the device default country.
        setupDeviceCountry()
    }

    //when a country is selected by the user, its position is sent back to the main activity.
    private fun countrySelected(position: Int) {

        val intent = Intent()
        intent.putExtra("country", position)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    private fun setupDeviceCountry() {

        val deviceCountry = deviceDefaultCountry

        //device country will be null if we are unable to find a match for the device's
        // selected country in our list of countries.
        if (deviceCountry == null) {
            binding.deviceDefault.root.visibility = View.GONE
            binding.defaultHeader.visibility = View.GONE
            return
        }

        //setup default country display
        val name = deviceCountry.name + ", " + deviceCountry.isoCode
        val dialingCode = deviceCountry.isdCode
        val referenceId: Int = deviceCountry.flagResource

        binding.deviceDefault.countryName.text = name
        binding.deviceDefault.countryCode.text = dialingCode
        binding.deviceDefault.flag.setImageResource(referenceId)

        binding.deviceDefault.root.setOnClickListener { countrySelected(-1) }
    }

}